oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/IndexWriterPool.java

```diff
@@ -0,0 +1,407 @@
+         */
+        CloseWriterOperation(LuceneIndexWriter delegate, long timestamp, SynchronousQueue<CloseResult> sync) {
+            super(delegate);
+            this.timestamp = timestamp;
+            this.sync = sync;
+        }
+
+        @Override
+        public void execute() {
+            try {
+                try {
+                    boolean closeResult = delegate.close(timestamp);
+                    sync.put(new CloseResult(closeResult));
+                } catch (IOException e) {
+                    sync.put(new CloseResult(e));
+                }
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    // Sentinel used to terminate worker threads
+    final static OperationBatch SHUTDOWN = new OperationBatch(-1, new Operation[0]);
+
+    private class Worker implements Runnable {
+        private final long startTime = System.nanoTime();
+        private final int id;
+        private long opCount = 0;
+        private long batchesProcessed = 0;
+        private long totalBusyTime = 0;
+
+        public Worker(int id) {
+            this.id = id;
+        }
+
+        @Override
+        public void run() {
+            String oldName = Thread.currentThread().getName();
+            Thread.currentThread().setName("index-writer-" + id);
+            try {
+                LOG.info("[{}] Worker started", id);
+                while (true) {
+                    OperationBatch op = queue.take();
+                    if (op == SHUTDOWN) {
+                        queue.add(SHUTDOWN);
+                        LOG.info("[{}] Shutting down worker", id);
+                        printStatistics();
+                        return;
+                    }
+                    long start = System.nanoTime();
+                    for (Operation operation : op.operations) {
+                        operation.execute();
+                        opCount++;
+                    }
+                    batchesProcessed++;
+                    long durationNanos = System.nanoTime() - start;
+                    totalBusyTime += durationNanos;
+                    long durationMillis = durationNanos / 1_000_000;
+                    if (durationMillis > 1000) {
+                        LOG.info("[{}] Processing batch {} of size {} took {} millis.",
+                                id, op.sequenceNumber, op.operations.length, durationMillis);
+                    }
+                    synchronized (pendingBatchesLock) {
+                        pendingBatches.remove(op.sequenceNumber);
+                        pendingBatchesLock.notifyAll();
+                    }
+                }
+            } catch (InterruptedException | IOException e) {
+                LOG.warn("[{}] Interrupted while waiting to take an update operation from the queue", id, e);
+                throw new RuntimeException(e);
```

Comment:
Here, I assume it makes sense to throw a RuntimeException

Reply:
Yes, here we are not trying to return a value to the thread that is waiting for these tasks to complete, the only information that needs to be sent back is success vs failure.